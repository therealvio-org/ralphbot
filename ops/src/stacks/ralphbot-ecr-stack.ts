import { Construct } from "constructs"
import { Stack, StackProps } from "aws-cdk-lib"
import { aws_ecr as ecr } from "aws-cdk-lib"

export class RalphbotEcrStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    new ecr.Repository(this, "ralphbotECRRepository", {
      lifecycleRules: [
        {
          maxImageCount: 5,
        },
      ],
      repositoryName: "ralphbot/master",
    })
  }
}
