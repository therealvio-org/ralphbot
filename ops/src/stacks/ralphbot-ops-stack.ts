import * as ec2 from "aws-cdk-lib/aws-ec2"
import * as ecr from "aws-cdk-lib/aws-ecr"
import * as ecs from "aws-cdk-lib/aws-ecs"
import * as iam from "aws-cdk-lib/aws-iam"
import * as logs from "aws-cdk-lib/aws-logs"
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager"
import * as ssm from "aws-cdk-lib/aws-ssm"
import { AppExtensionProps } from "../app"
import { Construct } from "constructs"
import { Stack, StackProps } from "aws-cdk-lib/core"

export class RalphbotStack extends Stack {
  constructor(
    scope: Construct,
    id: string,
    props?: StackProps,
    extendedProps?: AppExtensionProps
  ) {
    super(scope, id, props)

    const vpc = new ec2.Vpc(this, "vpc", {
      maxAzs: 1,
      natGateways: 0,
    })

    const repositoryRef = ecr.Repository.fromRepositoryArn(
      this,
      "ralphbotECRRepositoryRef",
      `arn:aws:ecr:${props?.env?.region}:${props?.env?.account}:repository/ralphbot/master`
    )

    const ralphbotImage = ecs.ContainerImage.fromEcrRepository(
      repositoryRef,
      extendedProps?.version
    )

    const cluster = new ecs.Cluster(this, "Cluster", {
      vpc: vpc,
    })
    const taskDefinition = new ecs.FargateTaskDefinition(
      this,
      "TaskDefinition",
      {
        cpu: 256,
        memoryLimitMiB: 512,
        runtimePlatform: {
          cpuArchitecture: ecs.CpuArchitecture.ARM64,
        },
      }
    )

    const botToken = secretsmanager.Secret.fromSecretNameV2(
      this,
      "botTokenFromName",
      "ralphbot/token"
    )
    const secretArnSuffix = ssm.StringParameter.valueForStringParameter(
      this,
      "/ralphbot/token/arn-suffix"
    )

    taskDefinition.addToExecutionRolePolicy(
      new iam.PolicyStatement({
        actions: [
          "secretsmanager:GetSecretValue",
          "secretsmanager:DescribeSecret",
        ],
        resources: [`${botToken.secretArn}-${secretArnSuffix}`],
      })
    )

    taskDefinition.addContainer("ralphbotContainer", {
      image: ralphbotImage,
      environment: {
        // trigger removal of registered commands on bot startup
        // this is configured to true on the app code itself, we're surfacing it here so it remains unobscured
        REMOVE_COMMANDS: "true",
      },
      secrets: {
        BOT_TOKEN: ecs.Secret.fromSecretsManager(botToken, "token"),
      },
      logging: ecs.LogDriver.awsLogs({
        streamPrefix: "ralphbot",
        logRetention: logs.RetentionDays.ONE_WEEK,
      }),
    })
    new ecs.FargateService(this, "Service", {
      assignPublicIp: true,
      cluster: cluster,
      taskDefinition: taskDefinition,
      circuitBreaker: { rollback: true },
      vpcSubnets: vpc.selectSubnets({
        subnetType: ec2.SubnetType.PUBLIC,
      }),
    })
  }
}
