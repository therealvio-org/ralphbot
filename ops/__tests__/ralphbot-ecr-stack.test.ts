import * as cdk from "aws-cdk-lib/core"
import { RalphbotEcrStack } from "../src/stacks/ralphbot-ecr-stack"
import { Template } from "aws-cdk-lib/assertions"

describe("Retrieve ralphbot ECR Stack", () => {
  const app = new cdk.App()
  const projectName = `ralphbot`

  const stack = new RalphbotEcrStack(app, `${projectName}-ecr`)

  const template = Template.fromStack(stack)

  it("Synths and snapshots", () => {
    expect(template).toMatchSnapshot()
  })
})
