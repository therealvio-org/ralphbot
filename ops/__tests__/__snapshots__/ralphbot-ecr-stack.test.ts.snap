// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Retrieve ralphbot ECR Stack Synths and snapshots 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ralphbotECRRepositoryF978983C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LifecyclePolicy": {
          "LifecyclePolicyText": "{"rules":[{"rulePriority":1,"selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":10},"action":{"type":"expire"}}]}",
        },
        "RepositoryName": "ralphbot/master",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
